main.go

```diff
@@ -1,3 +1,176 @@
 package main
 
-func main() {}
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"flag"
+	"io"
+	"io/fs"
+	"log"
+	"os"
+	"reflect"
+
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	apiyaml "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	crdDirectory = flag.String("dir", "", "")
+)
+
+func createClient() (client.Client, error) {
+	restConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	k8sClient, err := client.New(restConfig, client.Options{})
+	if err != nil {
+		return nil, err
+	}
+
+	return k8sClient, nil
+}
+
+func decodeCRDs(readCloser io.ReadCloser) ([]*v1.CustomResourceDefinition, error) {
+	reader := apiyaml.NewYAMLReader(bufio.NewReader(readCloser))
+	decoder := scheme.Codecs.UniversalDeserializer()
+
+	defer func(contentReader io.ReadCloser) {
+		err := readCloser.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(readCloser)
+
+	crdGVK := schema.GroupVersionKind{
+		Group:   "apiextensions.k8s.io",
+		Version: "v1",
+		Kind:    "CustomResourceDefinition",
+	}
+	var crds []*v1.CustomResourceDefinition
+	for {
+		doc, err := reader.Read()
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+
+		//  Skip over empty documents, i.e. a leading `---`
+		if len(bytes.TrimSpace(doc)) == 0 {
+			continue
+		}
+
+		var crd v1.CustomResourceDefinition
+		_, decodedGVK, err := decoder.Decode(doc, nil, &crd)
+		if err != nil {
+			return nil, err
+		} else if *decodedGVK != crdGVK {
+			continue
+		}
+		crds = append(crds, &crd)
+	}
+
+	return crds, nil
+}
+
+func findCRDs(filesystem fs.FS, dir string) ([]*v1.CustomResourceDefinition, error) {
+	var crds []*v1.CustomResourceDefinition
+	err := fs.WalkDir(filesystem, dir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+
+		documentReader, err := filesystem.Open(path)
+		if err != nil {
+			return err
+		}
+
+		documentCRDs, err := decodeCRDs(documentReader)
+		if err != nil {
+			return err
+		}
+
+		crds = append(crds, documentCRDs...)
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return crds, nil
+}
+
+func run() error {
+	filesystem := os.DirFS("./")
```

Comment:
I am not sure how `DirFS`works exactly but why not use  the `crdDirectory` here immediately?

Reply:
It's provides a local directory as the new fs.FS interface. This makes functions that touch the filesystem testable. It might be overkill here, but I was interested to try it.